File: TemperatureLogger.cs
using System.Globalization;
using System.Text;
using LibreHardwareMonitor.Hardware;

namespace EasyLogger;

public class TemperatureLogger : IDisposable
{
    private readonly int _interval;
    private readonly string _outputDir;
    private string? _logFilePath;
    private StreamWriter? _csvWriter;
    private Computer? _computer;
    private bool _disposed;

    public int LoggingInterval => _interval;
    public DateTime? LastReadingTime { get; private set; }

    public TemperatureLogger(int interval, string outputDir)
    {
        _interval = interval;
        _outputDir = outputDir;

        // Create output directory if it doesn't exist
        if (!Directory.Exists(_outputDir))
        {
            Directory.CreateDirectory(_outputDir);
        }
    }

    public void Initialize()
    {
        Console.WriteLine("[INFO] Initializing hardware monitoring...");

        try
        {
            // Initialize LibreHardwareMonitor Computer
            _computer = new Computer
            {
                IsCpuEnabled = true,
                IsGpuEnabled = true,
                IsMotherboardEnabled = true,
                IsMemoryEnabled = true,
                IsNetworkEnabled = false,
                IsStorageEnabled = true,
                IsControllerEnabled = true,
                IsPsuEnabled = true,
                IsBatteryEnabled = true
            };

            _computer.Open();
            Console.WriteLine("[OK] Hardware monitoring initialized");

            // Update hardware to get initial readings
            UpdateAllHardware();
        }
        catch (Exception ex)
        {
            Console.WriteLine("[ERROR] Failed to initialize hardware monitoring");
            Console.WriteLine($"  Make sure the application is running with administrator privileges");
            Console.WriteLine($"  Technical details: {ex.Message}");
            throw;
        }
    }

    private void UpdateAllHardware()
    {
        if (_computer == null) return;

        foreach (var hardware in _computer.Hardware)
        {
            hardware.Update();

            // Update sub-hardware (e.g., individual CPU cores)
            foreach (var subHardware in hardware.SubHardware)
            {
                subHardware.Update();
            }
        }
    }

    private void CreateLogFile()
    {
        var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
        _logFilePath = Path.Combine(_outputDir, $"temp_log_{timestamp}.csv");

        _csvWriter = new StreamWriter(_logFilePath, false, Encoding.UTF8);

        // Write CSV header
        _csvWriter.WriteLine("Timestamp,Hardware_Device,Sensor_Name,Sensor_Type,Value,Unit");
        _csvWriter.Flush();

        Console.WriteLine($"[OK] Created log file: {_logFilePath}");
    }

    private List<SensorReading> GetSensorData()
    {
        var readings = new List<SensorReading>();

        if (_computer == null) return readings;

        // Update all hardware before reading
        UpdateAllHardware();

        foreach (var hardware in _computer.Hardware)
        {
            CollectSensors(hardware, hardware.Name, readings);

            // Also collect from sub-hardware
            foreach (var subHardware in hardware.SubHardware)
            {
                CollectSensors(subHardware, hardware.Name, readings);
            }
        }

        return readings;
    }

    private void CollectSensors(IHardware hardware, string deviceName, List<SensorReading> readings)
    {
        foreach (var sensor in hardware.Sensors)
        {
            // Filter for relevant sensor types
            if (sensor.SensorType is SensorType.Temperature or
                                      SensorType.Load or
                                      SensorType.Clock or
                                      SensorType.Fan or
                                      SensorType.Power)
            {
                if (sensor.Value.HasValue)
                {
                    readings.Add(new SensorReading
                    {
                        DeviceName = deviceName,
                        SensorName = sensor.Name,
                        SensorType = sensor.SensorType.ToString(),
                        Value = sensor.Value.Value,
                        Unit = GetUnit(sensor.SensorType)
                    });
                }
            }
        }
    }

    private string GetUnit(SensorType sensorType)
    {
        return sensorType switch
        {
            SensorType.Temperature => "°C",
            SensorType.Load => "%",
            SensorType.Clock => "MHz",
            SensorType.Fan => "RPM",
            SensorType.Power => "W",
            _ => ""
        };
    }

    private void LogSensors()
    {
        if (_csvWriter == null) return;

        LastReadingTime = DateTime.Now;
        var timestamp = LastReadingTime.Value.ToString("yyyy-MM-dd HH:mm:ss");
        var sensors = GetSensorData();

        if (sensors.Count == 0)
        {
            Console.WriteLine($"[{timestamp}] WARNING: No sensor data available!");
            return;
        }

        // Write each sensor reading to CSV
        foreach (var sensor in sensors)
        {
            _csvWriter.WriteLine($"{timestamp},{EscapeCsv(sensor.DeviceName)},{EscapeCsv(sensor.SensorName)},{sensor.SensorType},{sensor.Value.ToString("F1", CultureInfo.InvariantCulture)},{sensor.Unit}");
        }

        _csvWriter.Flush();

        // Print summary to console
        var tempSensors = sensors.Where(s => s.SensorType == "Temperature").Take(5).ToList();
        if (tempSensors.Any())
        {
            var tempSummary = string.Join(", ", tempSensors.Select(s => $"{s.SensorName}: {s.Value:F1}°C"));
            Console.WriteLine($"[{timestamp}] Logged {sensors.Count} sensors ({sensors.Count(s => s.SensorType == "Temperature")} temperatures) - {tempSummary}");
        }
        else
        {
            Console.WriteLine($"[{timestamp}] Logged {sensors.Count} sensors");
        }
    }

    private string EscapeCsv(string value)
    {
        if (value.Contains(',') || value.Contains('"') || value.Contains('\n'))
        {
            return $"\"{value.Replace("\"", "\"\"")}\"";
        }
        return value;
    }

    public void Run()
    {
        Console.WriteLine();
        Console.WriteLine("======================================================================");
        Console.WriteLine("Temperature Logger Started");
        Console.WriteLine("======================================================================");
        Console.WriteLine($"Logging interval: {_interval} seconds");
        Console.WriteLine($"Output directory: {Path.GetFullPath(_outputDir)}");
        Console.WriteLine();
        Console.WriteLine("Press Ctrl+C to stop logging");
        Console.WriteLine();

        CreateLogFile();

        // Set up Ctrl+C handler
        Console.CancelKeyPress += (sender, e) =>
        {
            e.Cancel = true;
            Console.WriteLine();
            Console.WriteLine();
            Console.WriteLine("======================================================================");
            Console.WriteLine("Logging stopped by user");
            Console.WriteLine("======================================================================");
            Console.WriteLine($"Log file saved: {Path.GetFullPath(_logFilePath ?? "")}");
            Console.WriteLine();
            Environment.Exit(0);
        };

        try
        {
            while (true)
            {
                LogSensors();
                Thread.Sleep(_interval * 1000);
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[ERROR] {ex.Message}");
            throw;
        }
    }

    public void Dispose()
    {
        if (_disposed) return;

        _csvWriter?.Dispose();
        _computer?.Close();

        _disposed = true;
        GC.SuppressFinalize(this);
    }

    private class SensorReading
    {
        public string DeviceName { get; set; } = "";
        public string SensorName { get; set; } = "";
        public string SensorType { get; set; } = "";
        public float Value { get; set; }
        public string Unit { get; set; } = "";
    }
}

File: temperature_logger.py
"""
Temperature Logger for Windows
Monitors CPU, GPU, and motherboard temperatures using LibreHardwareMonitor
Logs data to CSV file for crash diagnosis
"""

import wmi
import csv
import time
import os
from datetime import datetime
import argparse
import configparser


class TemperatureLogger:
    def __init__(self, interval=10, output_dir="logs"):
        """
        Initialize the Temperature Logger

        Args:
            interval: Logging interval in seconds (default: 10)
            output_dir: Directory to save log files (default: "logs")
        """
        self.interval = interval
        self.output_dir = output_dir
        self.log_file = None
        self.csv_writer = None
        self.csv_file_handle = None

        # Ensure output directory exists
        if not os.path.exists(self.output_dir):
            os.makedirs(self.output_dir)

        # Initialize WMI connection
        try:
            self.w = wmi.WMI(namespace="root\\LibreHardwareMonitor")
            print("[OK] Connected to LibreHardwareMonitor")
        except Exception as e:
            print("[ERROR] Could not connect to LibreHardwareMonitor")
            print("  Make sure LibreHardwareMonitor is running with administrator privileges")
            print(f"  Technical details: {e}")
            raise

    def _create_log_file(self):
        """Create a new CSV log file with timestamp in filename"""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        self.log_file = os.path.join(self.output_dir, f"temp_log_{timestamp}.csv")

        self.csv_file_handle = open(self.log_file, 'w', newline='')
        self.csv_writer = csv.writer(self.csv_file_handle)

        # Write header
        self.csv_writer.writerow([
            "Timestamp",
            "Hardware_Device",
            "Sensor_Name",
            "Sensor_Type",
            "Value",
            "Unit"
        ])
        self.csv_file_handle.flush()

        print(f"[OK] Created log file: {self.log_file}")

    def _extract_hardware_name(self, parent_identifier):
        """
        Extract readable hardware name from parent identifier

        Args:
            parent_identifier: String like "/amdgpu/0" or "/intelcpu/0"

        Returns:
            Readable hardware name
        """
        if not parent_identifier:
            return "Unknown"

        # Try to get the hardware object to get its name
        try:
            hardware_objects = self.w.Hardware()
            for hw in hardware_objects:
                if hw.Identifier == parent_identifier:
                    return hw.Name
        except:
            pass

        # Fallback: parse identifier string
        # Remove leading/trailing slashes and split
        parts = parent_identifier.strip('/').split('/')
        if len(parts) > 0:
            hw_type = parts[0]
            # Convert common hardware types to readable names
            type_map = {
                'amdgpu': 'AMD GPU',
                'nvidiagpu': 'NVIDIA GPU',
                'intelcpu': 'Intel CPU',
                'amdcpu': 'AMD CPU',
                'mainboard': 'Motherboard',
                'ram': 'Memory',
                'hdd': 'Hard Drive',
                'nvme': 'NVMe SSD'
            }
            return type_map.get(hw_type.lower(), hw_type.upper())

        return "Unknown"

    def get_sensor_data(self):
        """
        Read all temperature sensors from LibreHardwareMonitor

        Returns:
            List of sensor readings as dictionaries
        """
        sensors = []

        try:
            # Query all hardware sensors
            sensor_data = self.w.Sensor()

            for sensor in sensor_data:
                # We're interested in Temperature, Load, Clock, and Fan sensors
                # Filter for relevant sensor types
                if sensor.SensorType in ['Temperature', 'Load', 'Clock', 'Fan', 'Power']:
                    sensors.append({
                        'name': sensor.Name,
                        'type': sensor.SensorType,
                        'value': sensor.Value,
                        'identifier': sensor.Identifier,
                        'parent': sensor.Parent
                    })

        except Exception as e:
            print(f"Warning: Error reading sensors: {e}")

        return sensors

    def log_sensors(self):
        """Read sensors and write to CSV file"""
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        sensors = self.get_sensor_data()

        if not sensors:
            print(f"[{timestamp}] WARNING: No sensor data available!")
            return

        # Write each sensor reading to CSV
        for sensor in sensors:
            # Determine unit based on sensor type
            unit = ""
            if sensor['type'] == 'Temperature':
                unit = "°C"
            elif sensor['type'] == 'Load':
                unit = "%"
            elif sensor['type'] == 'Clock':
                unit = "MHz"
            elif sensor['type'] == 'Fan':
                unit = "RPM"
            elif sensor['type'] == 'Power':
                unit = "W"

            # Extract hardware device name from parent identifier
            # Parent format is like: "/amdgpu/0" or "/intelcpu/0"
            hardware_device = self._extract_hardware_name(sensor['parent'])

            self.csv_writer.writerow([
                timestamp,
                hardware_device,
                sensor['name'],
                sensor['type'],
                sensor['value'],
                unit
            ])

        self.csv_file_handle.flush()

        # Print summary to console
        temp_sensors = [s for s in sensors if s['type'] == 'Temperature']
        if temp_sensors:
            temp_summary = ", ".join([f"{s['name']}: {s['value']:.1f}°C" for s in temp_sensors[:5]])
            print(f"[{timestamp}] Logged {len(sensors)} sensors ({len(temp_sensors)} temperatures) - {temp_summary}")
        else:
            print(f"[{timestamp}] Logged {len(sensors)} sensors")

    def run(self):
        """Main logging loop"""
        print("\n" + "="*70)
        print("Temperature Logger Started")
        print("="*70)
        print(f"Logging interval: {self.interval} seconds")
        print(f"Output directory: {os.path.abspath(self.output_dir)}")
        print("\nPress Ctrl+C to stop logging\n")

        self._create_log_file()

        try:
            while True:
                self.log_sensors()
                time.sleep(self.interval)

        except KeyboardInterrupt:
            print("\n\n" + "="*70)
            print("Logging stopped by user")
            print("="*70)
            print(f"Log file saved: {os.path.abspath(self.log_file)}")
            print()

        finally:
            if self.csv_file_handle:
                self.csv_file_handle.close()


def load_config(config_file='config.ini'):
    """
    Load configuration from config.ini file

    Returns:
        dict: Configuration values with 'interval' and 'output_dir'
    """
    config = configparser.ConfigParser()
    defaults = {
        'interval': 10,
        'output_dir': 'logs'
    }

    if os.path.exists(config_file):
        try:
            config.read(config_file)
            if 'Settings' in config:
                interval = config.getint('Settings', 'logging_interval', fallback=defaults['interval'])
                output_dir = config.get('Settings', 'output_directory', fallback=defaults['output_dir'])
                return {'interval': interval, 'output_dir': output_dir}
        except Exception as e:
            print(f"Warning: Could not read config file '{config_file}': {e}")
            print("Using default settings")

    return defaults


def main():
    # Load settings from config file
    config = load_config()

    parser = argparse.ArgumentParser(
        description='Temperature Logger for Windows using LibreHardwareMonitor',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python temperature_logger.py                    # Use settings from config.ini
  python temperature_logger.py -i 5               # Override interval to 5 seconds
  python temperature_logger.py -i 30 -o my_logs   # Override both interval and output directory

Configuration:
  Settings can be configured in config.ini file or overridden with command-line arguments.
  Command-line arguments take precedence over config.ini settings.
        """
    )

    parser.add_argument(
        '-i', '--interval',
        type=int,
        default=None,
        help=f'Logging interval in seconds (default from config.ini: {config["interval"]})'
    )

    parser.add_argument(
        '-o', '--output',
        type=str,
        default=None,
        help=f'Output directory for log files (default from config.ini: {config["output_dir"]})'
    )

    args = parser.parse_args()

    # Use command-line arguments if provided, otherwise use config file values
    interval = args.interval if args.interval is not None else config['interval']
    output_dir = args.output if args.output is not None else config['output_dir']

    # Validate interval
    if interval < 1:
        print("Error: Interval must be at least 1 second")
        return

    try:
        logger = TemperatureLogger(interval=interval, output_dir=output_dir)
        logger.run()
    except Exception as e:
        print(f"\nFatal error: {e}")
        return 1


if __name__ == "__main__":
    main()
